chainer/testing/test.py

```diff
@@ -0,0 +1,180 @@
+
+
+def _check_variable_types(vars, device, func_name, test_error_cls):
+    assert issubclass(test_error_cls, TestError)
+
+    if not isinstance(vars, tuple):
+        test_error_cls.fail(
+            '`{}()` must return a tuple, '
+            'not {}.'.format(func_name, type(vars)))
+    if not all(isinstance(a, chainer.Variable) for a in vars):
+        test_error_cls.fail(
+            '{}() must return a tuple of Variables.\n'
+            'Actual: {}'.format(
+                func_name, ', '.join(str(type(a)) for a in vars)))
+    if not all(isinstance(a.array, device.supported_array_types)
+               for a in vars):
+        test_error_cls.fail(
+            '{}() must return a tuple of Variables of arrays supported by '
+            'device {}.\n'
+            'Actual: {}'.format(
+                func_name, device, ', '.join(type(a.array) for a in vars)))
+
+
+def _check_forward_output_arrays_equal(
+        expected_arrays, actual_arrays, func_name, test_error_cls, **opts):
+    # `opts` is passed through to `testing.assert_all_close`.
+    # Check all outputs are equal to expected values
+    assert issubclass(test_error_cls, TestError)
+
+    message = None
+    while True:
+        # Check number of arrays
+        if len(expected_arrays) != len(actual_arrays):
+            message = (
+                'Number of outputs of {}() ({}, {}) does not '
+                'match'.format(
+                    func_name, len(expected_arrays), len(actual_arrays)))
+            break
+
+        # Check dtypes and shapes
+        dtypes_match = all([
+            ye.dtype == y.dtype
+            for ye, y in zip(expected_arrays, actual_arrays)])
+        shapes_match = all([
+            ye.shape == y.shape
+            for ye, y in zip(expected_arrays, actual_arrays)])
+        if not (shapes_match and dtypes_match):
+            message = (
+                'Shapes and/or dtypes of {}() do not match'.format(func_name))
+            break
+
+        # Check values
+        indices = []
+        for i, (expected, actual) in (
+                enumerate(zip(expected_arrays, actual_arrays))):
+            try:
+                array_module.assert_allclose(expected, actual, **opts)
+            except AssertionError:
+                indices.append(i)
+        if len(indices) > 0:
+            message = (
+                'Outputs of {}() do not match the expected values.\n'
+                'Indices of outputs that do not match: {}'.format(
+                    func_name, ', '.join(str(i) for i in indices)))
+            break
+        break
+
+    if message is not None:
+        test_error_cls.fail(
+            '{}\n'
+            'Expected shapes and dtypes: {}\n'
+            'Actual shapes and dtypes:   {}\n'.format(
+                message,
+                utils._format_array_props(expected_arrays),
+                utils._format_array_props(actual_arrays)))
+
+
+# Utilities for contiguousness tests.
```

Comment:
These should not be here: these are closely coupled with the tests. It's better to put within `tests` directory.

Reply:
Fixed.